File: agent/packages/schema/src/woliu_v2.ts
import { type Static, Type } from "@sinclair/typebox";

import { type ValidationResult, validate } from "./validate.js";

export const WoliuSkillIdV1 = Type.Union([
  Type.Literal("hold"),
  Type.Literal("burst"),
  Type.Literal("mouth"),
  Type.Literal("pull"),
  Type.Literal("heart"),
  Type.Literal("vacuum_palm"),
  Type.Literal("vortex_shield"),
  Type.Literal("vacuum_lock"),
  Type.Literal("vortex_resonance"),
  Type.Literal("turbulence_burst"),
]);
export type WoliuSkillIdV1 = Static<typeof WoliuSkillIdV1>;

export const WoliuBackfireLevelV1 = Type.Union([
  Type.Literal("sensation"),
  Type.Literal("micro_tear"),
  Type.Literal("torn"),
  Type.Literal("severed"),
]);
export type WoliuBackfireLevelV1 = Static<typeof WoliuBackfireLevelV1>;

export const WoliuSkillCastV1 = Type.Object(
  {
    caster: Type.String({ minLength: 1 }),
    skill: WoliuSkillIdV1,
    tick: Type.Integer({ minimum: 0 }),
    lethal_radius: Type.Number({ minimum: 0 }),
    influence_radius: Type.Number({ minimum: 0 }),
    turbulence_radius: Type.Number({ minimum: 0 }),
    absorbed_qi: Type.Number({ minimum: 0 }),
    swirl_qi: Type.Number({ minimum: 0 }),
    animation_id: Type.String({ minLength: 1 }),
    particle_id: Type.String({ minLength: 1 }),
    sound_recipe_id: Type.String({ minLength: 1 }),
    icon_texture: Type.String({ minLength: 1 }),
  },
  { additionalProperties: false },
);
export type WoliuSkillCastV1 = Static<typeof WoliuSkillCastV1>;

export const WoliuBackfireV1 = Type.Object(
  {
    caster: Type.String({ minLength: 1 }),
    skill: WoliuSkillIdV1,
    level: WoliuBackfireLevelV1,
    cause: Type.String({ minLength: 1 }),
    overflow_qi: Type.Number({ minimum: 0 }),
    tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type WoliuBackfireV1 = Static<typeof WoliuBackfireV1>;

export const TurbulenceFieldV1 = Type.Object(
  {
    caster: Type.String({ minLength: 1 }),
    skill: WoliuSkillIdV1,
    center: Type.Tuple([Type.Number(), Type.Number(), Type.Number()]),
    radius: Type.Number({ minimum: 0 }),
    intensity: Type.Number({ minimum: 0, maximum: 1 }),
    swirl_qi: Type.Number({ minimum: 0 }),
    tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type TurbulenceFieldV1 = Static<typeof TurbulenceFieldV1>;

export const WoliuPullDisplaceV1 = Type.Object(
  {
    caster: Type.String({ minLength: 1 }),
    target: Type.String({ minLength: 1 }),
    displacement_blocks: Type.Number({ minimum: 0 }),
    tick: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type WoliuPullDisplaceV1 = Static<typeof WoliuPullDisplaceV1>;

export function validateWoliuSkillCastV1Contract(data: unknown): ValidationResult {
  return validate(WoliuSkillCastV1, data);
}

export function validateWoliuBackfireV1Contract(data: unknown): ValidationResult {
  return validate(WoliuBackfireV1, data);
}

export function validateTurbulenceFieldV1Contract(data: unknown): ValidationResult {
  return validate(TurbulenceFieldV1, data);
}

export function validateWoliuPullDisplaceV1Contract(data: unknown): ValidationResult {
  return validate(WoliuPullDisplaceV1, data);
}

File: server/src/schema/woliu_v2.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WoliuSkillIdV1 {
    Hold,
    Burst,
    Mouth,
    Pull,
    Heart,
    VacuumPalm,
    VortexShield,
    VacuumLock,
    VortexResonance,
    TurbulenceBurst,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WoliuBackfireLevelV1 {
    Sensation,
    MicroTear,
    Torn,
    Severed,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WoliuSkillCastV1 {
    pub caster: String,
    pub skill: WoliuSkillIdV1,
    pub tick: u64,
    pub lethal_radius: f32,
    pub influence_radius: f32,
    pub turbulence_radius: f32,
    pub absorbed_qi: f32,
    pub swirl_qi: f32,
    pub animation_id: String,
    pub particle_id: String,
    pub sound_recipe_id: String,
    pub icon_texture: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WoliuBackfireV1 {
    pub caster: String,
    pub skill: WoliuSkillIdV1,
    pub level: WoliuBackfireLevelV1,
    pub cause: String,
    pub overflow_qi: f32,
    pub tick: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TurbulenceFieldV1 {
    pub caster: String,
    pub skill: WoliuSkillIdV1,
    pub center: [f64; 3],
    pub radius: f32,
    pub intensity: f32,
    pub swirl_qi: f32,
    pub tick: u64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WoliuPullDisplaceV1 {
    pub caster: String,
    pub target: String,
    pub displacement_blocks: f32,
    pub tick: u64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn woliu_skill_cast_serializes_snake_case_skill() {
        let payload = WoliuSkillCastV1 {
            caster: "player:a".to_string(),
            skill: WoliuSkillIdV1::VacuumPalm,
            tick: 7,
            lethal_radius: 3.0,
            influence_radius: 30.0,
            turbulence_radius: 10.0,
            absorbed_qi: 0.1,
            swirl_qi: 12.0,
            animation_id: "bong:vortex_palm_open".to_string(),
            particle_id: "bong:vortex_spiral".to_string(),
            sound_recipe_id: "vortex_qi_siphon".to_string(),
            icon_texture: "bong:textures/gui/skill/woliu_mouth.png".to_string(),
        };
        let json = serde_json::to_string(&payload).unwrap();
        assert!(json.contains(r#""skill":"vacuum_palm""#));
        assert!(json.contains("vortex_qi_siphon"));
    }
}
